pkg/operator/staticpod/controller/guard/guard_controller.go

```diff
@@ -361,5 +371,11 @@ func (c *GuardController) sync(ctx context.Context, syncCtx factory.SyncContext)
 		}
 	}
 
+	// For cases which take longer or never to reconcile wait for some time
+	// to avoid quick reconciliation
+	if errWithDelay {
+		time.Sleep(c.delayTimeout)
+	}
```

Comment:
Two comments about that delay.

1. reconciliation should be asynchronous, so I would expect the controller to continue to reconcile regardless of the state of the previous sync, but I might be wrong
2. adding delay here, will delay the report of the degraded status from https://github.com/openshift/library-go/pull/1624/files#diff-93e47de48c603f6e2ddfe947c96610f13116e244dd7127845f8d23ea3d244c0dR112

I wonder if we have some precedence in library-go of doing something like that, usually I would expect having a work queue with some backoff

Reply:
yes, I agree with @dgrisonnet , usually the solution here is to add the key to the queue asynchronously with a delay. (i think work queue has support for this)
But the issue is, our controller is not key based, it is periodic (like every 1s or something)

Reply:
The guard controller's event handlers registers changes from pods under openshift-kube-scheduler and from nodes. When saturating all master nodes with system-node-critical priority class pods the guard pods start to be periodically terminated and re-created. For 3 master nodes cluster and with the priority and fairness enabled it takes approximately 10 seconds to run the sync method. In these 10 seconds the event handler for pods is invoked multiple times which marks the "key" item in the queue as dirty. Meaning, even when the "key" item is popped from the queue, it never gets processed without getting repeatedly enqueued back. Thus, the "key" item never leaves the queue for more than a second or two.